README.md
mythfs-perl
===========

This is a FUSE filesystem for MythTV (www.mythtv.org).  It uses the
Myth 0.25 API to mount the TV recordings known to a MythTV master
backend onto a virtual filesystem on the client machine for convenient
playback with mplayer or other video tools. Because it uses the MythTV
network protocol, the recordings do not need to be on a shared
NFS-mounted disk, nor does the Myth database need to be accessible
from the client.

Installation
============

Run the following commands from within the top-level directory of this
distribution:

 <pre> 
 $ <b> git clone git://github.com/lstein/mythfs-perl.git
 $ <b>cd mythfs-perl</b>
 $ <b>./Build.PL</b>
 $ <b>./Build test</b>
 $ <b>sudo ./Build install</b>
</pre>

If you get messages about missing dependencies, run:

<pre>
 $ <b>./Build installdeps</b>
</pre>

and then "sudo ./Build install". See Fuse Notes if you get
Fuse-related errors when attempting to install dependencies.

For best performance, Perl must have been compiled with support for
IThreads. In addition, you will need at Fuse version 0.15 to run
correctly under Perl version >= 5.14.  Threading will be automatically
disabled if not available. See Fuse Notes for additional information.

Usage
=====

To mount the recordings contained on the master backend "MyHost" onto
a local filesystem named "/tmp/mythfs" use this command:

<pre>
 $ <b>mkdir /tmp/mythfs</b>
 $ <b>mythfs.pl MyHost /tmp/mythfs</b>
</pre>

The script will fork into the background and should be stopped with
fusermount. The mounted /tmp/mythfs directory will contain a series of
human-readable recordings organized by title (directory) and subtitle
(file). 

To unmount:

<pre>
 $ <b>fusermount -u /tmp/mythfs</b>
</pre>

NOTE: Do NOT try to kill the mythfs.pl process. This will only cause a
hung filesystem that needs to be unmounted with fusermount.

There are a number of options that you can pass to mythfs.pl,
including the ability to customize the filesystem layout and set the
interval that the backend is checked for new and deleted
recordings. Call mythfs.pl with the -h option for the complete help
text.

Local Recordings
================

The default behavior of this filesystem is to use the Myth API to
stream recordings across the network when you attempt to read from
them. This is done in an efficient way that fetches just the portion
of the file you wish to read. However, if the underlying recording
files are directly accessible (either in a regular director or via an
NFS mount), you can get better performance by passing mythfs.pl the
--mountpt option with the path to the directory in which the
recordings can be found. The filesystem will then be set up as a set
of symbolic links that point from a human readable file name to the
recording file.

The main advantage of creating symbolic links is that NFSv4 can be
significantly faster than the backend streaming protocol -- about 1.6X
in my informal tests. The main limitation is that this mode does not
understand storage groups, so all recordings need to be located in a
single storage group in a locally-accessible directory. However if a
recording file is not found in local directory, then mythfs.pl will
fall back to the streaming protocol, so the recording is accessible
one way or another.

The Default Directory Layout
============================

Recordings that are part of a series usually have a title (the series
name) and subtitle (the episode name). Such recordings are displayed
using a two-tier directory structure in which the top-level directory
is the series name, and the contents are a series of recorded
episodes. The corresponding pattern (as described in the next section)
is "%T/%S".

For recordings that do not have a subtitle, typically one-off movie
showings, the recording is placed at the top level.

If needed for uniqueness, the channel number and time the recorded was
started is attached to the filename, along with an extension
indicating the recording type (.mpg or .nuv). The file create and
modification times correspond to the recording start time. For
directories, the times are set to the most recent recording contained
within the directory.

Here is an example directory listing:

<pre>
 % <b>ls -lR  /tmp/mythfs</b>
 total 35
 -r--r--r-- 1 lstein lstein 12298756208 Dec 30 00:00 A Funny Thing Happened on the Way to the Forum.mpg
 -r--r--r-- 1 lstein lstein 14172577964 Dec 25 16:00 A Heartland Christmas.mpg
 dr-xr-xr-x 1 lstein lstein           5 Mar 11 03:00 Alfred Hitchcock Presents
 dr-xr-xr-x 1 lstein lstein           8 May  2 00:00 American Dad
 ...

 /home/lstein/Myth/Alfred Hitchcock Presents:
 total 3
 -r--r--r-- 1 lstein lstein 647625408 Dec 25 15:30 Back for Christmas.mpg
 -r--r--r-- 1 lstein lstein 647090360 Dec  7 00:00 Dead Weight.mpg
 -r--r--r-- 1 lstein lstein 660841056 Mar 11 03:00 Rose Garden.mpg
 -r--r--r-- 1 lstein lstein 647524452 Dec 25 00:00 Santa Claus and the 10th Ave. Kid.mpg
 -r--r--r-- 1 lstein lstein 649819932 Dec 27 00:00 The Contest of Aaron Gold.mpg

 /home/lstein/Myth/American Dad:
 total 4
 -r--r--r-- 1 lstein lstein 3512038152 Apr 24 00:00 Flirting With Disaster.mpg
</pre>

The size of directories corresponds to the number of recordings (not
counting subdirectories) contained within it. The modification time of
directories is the start time of the most recent recording contained
within it.

Customizing the Directory Listing
=================================

You may customize the directory listing by providing a pattern for
naming each recording using the -p option. For example:

 $ mythfs.pl -p '%C/%T:%S (%od-%ob-%oY)' mythbackend ~/Myth

This will create filenames that look like this:

<pre>
 Sitcom/The Simpsons:The Food Wife (13-Nov-2011).mpg
</pre>

Patterns contain a combination of constant strings plus substitution
patterns consisting of the "%" sign plus 1 to three characters. A
slash will be interpreted as a directory level: multiple levels are
allowed. 

Commonly-used substitution patterns are:

    %T   = Title (show name)
    %S   = Subtitle (episode name)
    %C   = Category
    %cn  = Channel: channel number
    %cN  = Channel: channel name
    %y   = Recording start time:  year, 2 digits
    %Y   = Recording start time:  year, 4 digits
    %m   = Recording start time:  month, leading zero
    %b   = Recording start time:  abbreviated month name
    %B   = Recording start time:  full month name
    %d   = Recording start time:  day of month, leading zero
    %h   = Recording start time:  12-hour hour, with leading zero
    %H   = Recording start time:  24-hour hour, with leading zero
    %i   = Recording start time:  minutes
    %s   = Recording start time:  seconds
    %a   = Recording start time:  am/pm
    %A   = Recording start time:  AM/PM

A full list of patterns can be obtained by running "mythfs.pl -p
help".

Patterns are largely compatible with the excellent mythlink.pl
(http://www.mythtv.org/wiki/Mythlink.pl) script, but there are a small
number of enhancements, such as the ability to generate the month
name. Also, the patterns that generate the month name without a
leading zero are not supported.

You may wish to use a delimiter to separate fields of the recording
name, for example "%T:%S" to generate "Title:Subtitle". Occasionally a
recording field is empty, leading to names like "The Wild
Ones:.mpg". To avoid this, pass the --trim option with the delimiter
you use, and dangling/extra delimiters will be trimmed:

<pre>
 $ mythfs.pl -p '%T:%S' --trim=':' backend /tmp/myth
</pre>

If after applying the pattern to a recording the resulting path is not
unique, then this script will uniqueify the path by appending to it
the channel number and recording start time, for example:

 Masterpiece Classic/Downtown Abbey_17_1-2013-02-11T02:00.mpg
 Masterpiece Classic/Downtown Abbey_17_1-2013-03-10T06:00.mpg

Caching
=======

New and updated recordings will appear in the filesystem after a
slight delay due to the manner in which the script caches the
recording list. By default the backend is only checked for updates
every 10 minutes, but you can adjust this using the --cachetime
option, which takes the interval in minutes at which the system
checks for new and updated recordings.

For example, this command will reduce the update interval to 2
minutes:

  $ <b>mythfs.pl MyHost --cachetime=2 /tmp/mythfs</b>

Fuse Notes
==========

For best performance, you will need to run this filesystem using a
version of Perl that supports IThreads. Otherwise it will fall back to
non-threaded mode, which will introduce occasional delays during
directory listings and have notably slower performance when reading
from more than one file simultaneously.

If you are running Perl 5.14 or higher, you *MUST* use at least 0.15
of the Perl Fuse module. At the time this was written, the version of
Fuse 0.15 on CPAN was failing its regression tests on many
platforms. I have found that the easiest way to get a fully
operational Fuse module is to clone and compile a patched version of
the source, following this recipe:

<pre>
 $ <b>git clone git://github.com/isync/perl-fuse.git</b>
 $ <b>cd perl-fuse</b>
 $ <b>perl Makefile.PL</b>
 $ <b>make test</b>   (optional)
 $ <b>sudo make install</b>
</pre>


Troubleshooting
===============

This script has not yet undergone diligent testing. Try running with
the -debug flag to see where the problems are occurring and report
issues to https://github.com/lstein/mythfs-perl.

Author
======

Copyright 2013, Lincoln D. Stein <lincoln.stein@gmail.com>

License
=======

This package is distributed under the terms of the Perl Artistic
License 2.0. See http://www.perlfoundation.org/artistic_license_2_0.


Changes
1.35 Sat Jul  1 15:58:56 EDT 2017

    * Added a %PLX substitution pattern to create a hierarchy of
      directories that follows Plex Media Service's filenaming
      conventions
    * Fixed bug in processing of the %se (Season) and %e (Episode)
      substitution patterns.
    
1.34 Thu May 16 07:09:12 EDT 2013
    * Create an UPCOMING file to list upcoming recordings.
    * unlinking files and directories (e.g. using "rm") will now
      delete the underlying recording from the backend.

1.33 Mon May 13 12:38:20 EDT 2013
     * Significantly improved performance (~30%) by using LWP and activating keepalives.

1.32 Sun May 12 12:12:24 EDT 2013
     * Create STATUS file at top level, and better reporting of errors.
     * Add %TC and %ST substitution patterns (needs generalization).

1.31 Fri May 10 06:53:24 EDT 2013
     * Documentation formatting fixes.

1.30 Fri May 10 00:37:52 EDT 2013
     * Modularized the script into a series of libraries.

1.23 Thu May  9 15:26:28 EDT 2013
     * Replace WWW::Curl with HTTP::Lite due to thread-related crashes in Curl library.     

1.22 Wed May  8 23:27:17 EDT 2013
     * Always create an automatic file named .fuse-mythfs containing version and author information.
     * Added copyright and license information.

1.01-1.21
     * Stability fixes and documentation.

1.00 Mon May 6 2013
     * First public release



lib/Net/MythTV/Fuse.pm
package Net::MythTV::Fuse;

=head1 NAME

Net::MythTV::Fuse - Mount Fuse filesystem to display TV recordings managed by a MythTV backend

=head1 SYNOPSIS

 $filesystem = Net::MythTV::Fuse->new(
        mountpoint      => '/tmp/mythtv',
        backend         => 'mythbackend.domain.org',
       );

 $filesystem->run();

=head1 DESCRIPTION

This module uses Fuse to mount a virtual filesystem representing the TV
recordings captured by a local or remote MythTV backend
(www.mythtv.org). The backend must be at version 0.25 or higher.

Typically you will be using the frontend script, mythfs.pl rather than
this module directly.

=head1 METHODS

=cut


use strict;
use warnings;
use Net::MythTV::Fuse::Recordings;
use Fuse 'fuse_get_context';
use POSIX qw(ENOENT EISDIR ENOTDIR ENOTEMPTY EINVAL ECONNABORTED EACCES EIO);
use Data::Dumper 'Dumper'; # for DEBUGGING!
use Carp 'croak';

use constant MARKER_FILE=> '.fuse-mythfs';
use constant STATUS_FILE=> 'STATUS';
use constant UPCOMING_FILE=> 'UPCOMING';

our $VERSION = '1.35';

my $Package = __PACKAGE__;
my ($Recorded,$Options);

=head2 $f = Net::MythTV::Fuse->new(@options)

Create a new Fuse filesystem object according to the options. The
options are a set of key/value pairs and can be one or more of the
following:

    mountpoint           Mountpoint for the Fuse filesystem. Required and no default.
    backend              IP address of the backend (localhost)
    port                 Control port for the backend (6544)
    debug                Enable debugging if true (numeric; increasing values increase verbosity)
    threaded             Enable threading if true (true)
    fuse_options         Comma-delimited set of mount options to pass to Fuse, such as "allow_other" (none)
    cachetime            Time, in seconds, to cache list of recordings before refreshing from backend (300)
    pattern              Template for transforming recordings into paths (%T/%S)
    delimiter            Trim this string from the pathname if it is dangling or occurs multiple times (none)
    localmount           If storage group is locally (or NFS) mounted, use this mount point for direct access (none)
    dummy_data_path      Path to an XML-formatted list of recordings, used for debugging (none)

See the help text for mythfs.pl for more information on these arguments.

=cut

sub new {
    my $self     = shift;
    my %options  = @_;
    $options{mountpoint} && $options{backend}
        or croak "Usage: $self->new(mountpoint=>'\$mtpt',backend=>'\$backend,\@other_options)";

    my $recordings = Net::MythTV::Fuse::Recordings->new(\%options);
    $recordings->load_dummy_data($options{dummy_data_path})
	               if $options{dummy_data_path};

    return bless { options    => \%options,
		   recordings => $recordings
                 },ref $self || $self;
}

=head2 Read-only Accessors

The following are read-only accessors that return information about
the filesystem options:

 recordings     A Net::MythTV::Fuse::Recordings object used to fetch the recording list
 options        Hashref of options passed during object creation
 debug          Debug flag
 mountpoint     Virtual filesystem mount point
 localmount     Local mountpoint for recordings, if any
 threaded       True if threading enabled
 marker_file    Name of an automatically generated file (".fuse-mythfs") that will appear
                     at the top level of the virtual filesystem.

=cut

sub recordings {shift->{recordings}}
sub options    {shift->{options}}
sub debug      {shift->{options}{debug}}
sub mountpoint {shift->{options}{mountpoint}}
sub localmount {shift->{options}{localmount}}
sub threaded   {shift->{options}{threaded}}
sub marker_file { return MARKER_FILE }

=head2 $f->run()

This method will create the virtual filesystem and will not return
until the process is either killed, or the filesystem is unmounted
with fusermount -u.

=cut

sub run {
    my $self    = shift;
    my $options = $self->options;

    # copy critical parameters into package globals so that Fuse callbacks work
    $Recorded   = $self->recordings;
    $Options    = $options;

    $self->recordings->start_update_thread;
    
    Fuse::main(mountpoint => $options->{mountpoint},
	       getdir     => "$Package\:\:e_getdir",
	       getattr    => "$Package\:\:e_getattr",
	       open       => "$Package\:\:e_open",
	       unlink     => "$Package\:\:e_unlink",
	       rmdir      => "$Package\:\:e_rmdir",
	       read       => "$Package\:\:e_read",
	       release    => "$Package\:\:e_release",
	       readlink   => "$Package\:\:e_readlink",
	       mountopts  => $options->{fuse_options}||'',
	       debug      => $self->debug > 1,
	       threaded   => $self->threaded,
	);
}


=head2 Non-object methods

The following are ordinary subroutines that are used by Fuse to
generate the virtual filesystem:

 $clean_path = fixup($path)                Remove the leading '/' from file paths.
 $status     = e_open($path)               Check that $path can be opened and return 0 if so.
 $status     = e_release($path)            Called to release a closed path, returns 0 if successful.
 $contents   = e_read($path,$size,$offset) Called to read $size bytes from $path starting at $offset.
 @entries    = e_getdir($path)             Return all entries within directory indicated by $path.
 $contents   = e_readlink($path)           Resolve a symbolic link (used when recordings mounted locally).
 @attributes = e_getattr($path)            Stat() call on $path
 $status     = e_unlink($path)             Called to remove a file, returns 0 if successful.
 $status     = e_rmdir($path)              Called to remove a directory, returns 0 if successful.
 $string     = copyright_and_version()     Return contents of the automatic file ".fuse-mythfs"

=cut

sub fixup {
    my $path = shift;
    $path =~ s!^/!!;
    $path;
}

sub e_open {
    my $path = fixup(shift);

    # awkward: fix
    return 0 if $path eq MARKER_FILE or $path eq STATUS_FILE or $path eq UPCOMING_FILE;
    
    $Recorded->valid_path($path) or return -ENOENT();
    $Recorded->is_dir($path)    and return -EISDIR();
    return 0;
}

sub e_release {
    return 0;
}

sub e_unlink {
    my $path   = fixup(shift);

    $Recorded->valid_path($path) or return -ENOENT();
    $Recorded->is_dir($path)    and return -EISDIR();
    return -EACCES() if $path eq MARKER_FILE or $path eq STATUS_FILE or $path eq UPCOMING_FILE;   

    my $status = $Recorded->delete_recording($path);
    $status eq 'ok' or return -EIO();
    return 0;
}

sub e_rmdir {
    my $path   = fixup(shift);
    $Recorded->valid_path($path) or return -ENOENT();
    $Recorded->is_dir($path)     or return -ENOTDIR();
    my $status = $Recorded->delete_directory($path);
    $status eq 'directory not empty' and return -ENOTEMPTY();
    $status eq 'ok' or return -EIO();
    return 0;
}

sub e_read {
    my ($path,$size,$offset) = @_;
    $offset ||= 0;
    $path = fixup($path);

    # we need to create a special method for these special files
    if ($path eq MARKER_FILE) {
	my $content = copyright_and_version();
	return substr($content,$offset,$size);
    }

    if ($path eq STATUS_FILE) {
	my $content = $Recorded->status;
	return substr($content,$offset,$size);
    }

    if ($path eq UPCOMING_FILE) {
	my $content = _get_upcoming_list();
	return substr($content,$offset,$size);
    }

    my ($retcode,$contents) = $Recorded->download_recorded_file($path,$size,$offset);
    return -ENOENT()       if $retcode eq 'not found';
    return -EINVAL()       if $retcode eq 'invalid offset';
    return -ECONNABORTED() if $retcode eq 'connection failed';
    return $contents;
}

sub e_getdir {
    my $path = fixup(shift) || '.';

    my @entries = $Recorded->entries($path);
    unshift @entries,MARKER_FILE,STATUS_FILE,UPCOMING_FILE if $path eq '.';
    return -ENOENT() unless @entries;
    return ('.','..',@entries,0);
}

sub e_readlink {
    my $path = fixup(shift) || '.';
    my $basename   = $Recorded->basename($path) or return -ENOENT();
    my $localmount = $Recorded->localmount      or return -ENOENT();
    my $local_path = "$localmount/$basename";
    return $local_path;
}

sub e_getattr {
    my $path = fixup(shift) || '.';

    my $context = fuse_get_context();
    my ($dev, $ino, $rdev, $blocks, $gid, $uid, $nlink, $blksize) 
	= (0,0,0,1,@{$context}{'gid','uid'},1,1024);

    if ($path eq MARKER_FILE) { # special case
	my $contents = copyright_and_version();
	return (
	    $dev,$ino,0100000|0444,$nlink,$uid,$gid,$rdev,
	    length($contents),time(),time(),time(),$blksize,$blocks);
    }

    if ($path eq STATUS_FILE) { # another special case
	my $contents = $Recorded->status;
	return (
	    $dev,$ino,0100000|0444,$nlink,$uid,$gid,$rdev,
	    length($contents),time(),time(),time(),$blksize,$blocks);
    }

    if ($path eq UPCOMING_FILE) { # another special case
	my $content = _get_upcoming_list();
	return (
	    $dev,$ino,0100000|0444,$nlink,$uid,$gid,$rdev,
	    length($content),time(),time(),time(),$blksize,$blocks);
    }

    my $entry    = $Recorded->entry($path) or return -ENOENT();

    # debugging an occasional error
    unless ($entry->{type}) {
	warn "broken entry for path =$path";
	print STDERR Data::Dumper::Dumper($entry);
	return -ENOENT();
    }

    my $basename = $entry->{basename};
    my $isdir    = $entry->{type} eq 'directory';
    my $islink   = $entry->{type} eq 'file' && $Recorded->localmount && -r $Recorded->localmount."/$basename";

    my $readonly = $Options->{fuse_options}=~ /\bro\b/;
    my $write = $readonly ? 0000                : 0200;
    my $mode  = $isdir    ? 0040000|0555|$write : ($islink ? 0120000|0777 : 0100000|0444|$write);

    my $ctime = $entry->{ctime};
    my $mtime = $entry->{mtime};
    my $atime = $mtime;
    my $size  = $entry->{length};

    return ($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,
	    $size,$atime,$mtime,$ctime,$blksize,$blocks);
}

sub copyright_and_version {
    return <<END;
mythfs.pl version $VERSION. 
Copyright 2013 Lincoln D. Stein <lincoln.stein\@gmail.com>. 
Distributed under Perl Artistic License Version 2.
END
}

sub _get_upcoming_list {
    my $self = shift;
    return $Recorded->get_upcoming_list;
}


1;


=head1 AUTHOR

Copyright 2013, Lincoln D. Stein <lincoln.stein@gmail.com>

=head1 LICENSE

This package is distributed under the terms of the Perl Artistic
License 2.0. See http://www.perlfoundation.org/artistic_license_2_0.

=cut

__END__
